Source/MQTTnet/Server/Internal/MqttClientSubscriptionsManager.cs

```diff
@@ -172,11 +173,17 @@ public async Task<SubscribeResult> Subscribe(MqttSubscribePacket subscribePacket
                 throw new ArgumentNullException(nameof(subscribePacket));
             }
 
-            var retainedApplicationMessages = await _retainedMessagesManager.GetMessages().ConfigureAwait(false);
             var result = new SubscribeResult(subscribePacket.TopicFilters.Count);
 
             var addedSubscriptions = new List<string>();
             var finalTopicFilters = new List<MqttTopicFilter>();
+            var atLeastOnceSubscriptionResults = new List<CreateSubscriptionResult>();
+
+            IList<MqttApplicationMessage> retainedApplicationMessages = null;
+            if (subscribePacket.TopicFilters.Any(f => f.QualityOfServiceLevel != MqttQualityOfServiceLevel.AtLeastOnce))
```

Comment:
Why is this only relevant for at least once QoS? Isn't this also beneficial for QoS 0 and 2?
Or are you fixing a specific issue in your project here?

Reply:
Good question. The project wherein we discovered this bug uses solely `AtLeastOnce`. But I think you're right that it may fix potential issues with `AtMostOnce` and `ExactlyOnce` as well.

Reply:
@chkr1011 Is this PR sufficient as-is or are you requesting that I make changes?